### PR TITLE
feat: update priority fee estimator

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Current estimator version is not flexible and will always return 3 gwei priority fee if base fee is less than 100 gwei which is not correct for L2s (and often pretty high even for Mainnet).

## Solution

New impl was inspired by Metamask estimator and introduces following changes:
1. We no more use `EIP1559_FEE_ESTIMATION_THRESHOLD_MAX_CHANGE` because it could've resulted in overestimated gas prices. Let's consider `rewards` with following contents: `[100, 1, 1, 1, 1]`, in this case we'd filter out all 1's because of spike caused by a single 100 price, and the resulted priority fee estimate will be 100 gwei which is incorrect. Using median should be enough for a fair estimate filtering out potential anomalies.
2. We now simply multiply base fee by 2 to calculate max base fee instead of using `base_fee_surged` oriented on mainnet.

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

File: crates/provider/src/utils.rs [hunk 1]
```diff
@@ -23,75 +17,16 @@ fn estimate_priority_fee(rewards: Vec<Vec<U256>>) -> U256 {
...
-    if base_fee_per_gas <= U256::from(40_000_000_000u64) {
-        base_fee_per_gas * U256::from(2)
-    } else if base_fee_per_gas <= U256::from(100_000_000_000u64) {
-        base_fee_per_gas * U256::from(16) / U256::from(10)
-    } else if base_fee_per_gas <= U256::from(200_000_000_000u64) {
-        base_fee_per_gas * U256::from(14) / U256::from(10)
-    } else {
-        base_fee_per_gas * U256::from(12) / U256::from(10)
-    }
+    rewards[rewards.len() / 2]
```
Comment (mattsse) [inline crates/provider/src/utils.rs:23]: using the median by default is sane

File: crates/provider/src/utils.rs [hunk 2]
```diff
@@ -1,19 +1,13 @@
 //! Provider-related utilities.
 
-use alloy_primitives::{I256, U256};
+use alloy_primitives::U256;
 
 /// The number of blocks from the past for which the fee rewards are fetched for fee estimation.
 pub const EIP1559_FEE_ESTIMATION_PAST_BLOCKS: u64 = 10;
+/// Multiplier for the current base fee to estimate max base fee for the next block.
+pub const EIP1559_BASE_FEE_MULTIPLIER: f64 = 2.0;
```
Comment (mattsse) [inline crates/provider/src/utils.rs:8]: that should be more reasonable

**Review by mattsse — APPROVED** [review]
lgtm

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/provider/src/utils.rs (hunk 1 above)
Comment (DaniPopes) [inline crates/provider/src/utils.rs:23]:
```suggestion
    let n = rewards.len();
    if n % 2 == 0 {
        (rewards[n / 2] + rewards[n / 2 + 1]) / 2
    } else {
        rewards[(n + 1) / 2]
    }
```

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/provider/src/utils.rs (hunk 1 above)
Comment (klkvr) [reply crates/provider/src/utils.rs]: wouldn't this panic when n == 2?

Comment (DaniPopes) [reply crates/provider/src/utils.rs]: sure, just flagging this is not the real median

Comment (klkvr) [reply crates/provider/src/utils.rs]: 22e3f0df96b23ac1efde66fbe8ebe61cd50d1bd3

File: crates/provider/src/utils.rs [hunk 3]
```diff
@@ -1,97 +1,40 @@
...
-        potential_max_fee
-    };
-    (max_fee_per_gas, max_priority_fee_per_gas)
+/// The default EIP-1559 fee estimator which is based on the work by [MetaMask](https://github.com/MetaMask/core/blob/main/packages/gas-fee-controller/src/fetchGasEstimatesViaEthFeeHistory/calculateGasFeeEstimatesForPriorityLevels.ts#L56)
+/// (constants for "medium" priority level are used)
+pub fn eip1559_default_estimator(base_fee_per_gas: U256, rewards: &[Vec<U256>]) -> (U256, U256) {
+    let max_priority_fee_per_gas = estimate_priority_fee(rewards);
+    let potential_max_fee = base_fee_per_gas * U256::from(EIP1559_BASE_FEE_MULTIPLIER);
+    (potential_max_fee, max_priority_fee_per_gas)
 }
```
Comment (DaniPopes) [inline crates/provider/src/utils.rs:40]: Can we add some sanity tests? I don't see any

**Review by DaniPopes — APPROVED** [review]
Needs test otherwise LGTM

**Pull request merged**